trieste/acquisition/function.py

```diff
@@ -512,15 +501,158 @@ def prepare_acquisition_function(
         return lambda at: expected_improvement(objective_model, eta, at) * constraint_fn(at)
 
 
-class BatchAcquisitionFunctionBuilder(ABC):
-    """ A :class:`BatchAcquisitionFunctionBuilder` builds a batch acquisition function. """
+def independent_latent_sampler_with_reparametrisation(
+    dataset: Dataset, model: ProbabilisticModel, sample_size: int, observation_shape: ShapeLike,
+) -> Callable[[tf.Tensor], tf.Tensor]:
+    """
+    Returns independent samples according to the reparametrization trick.
+    The sample size S is determined by eps,
+    N is the number of query points and L is the number of latent processes.
+    :param dataset:
+    :param model:
+    :param sample_size:
+    :param observation_shape:
+    :return: a [S, N, L] tensor
+    """
+    eps = tf.random.normal([sample_size] + observation_shape, dtype=dataset.observations.dtype)
 
-    @abstractmethod
-    def prepare_acquisition_function(
-        self, datasets: Mapping[str, Dataset], models: Mapping[str, ProbabilisticModel]
-    ) -> BatchAcquisitionFunction:
+    def sample(at: tf.Tensor) -> tf.Tensor:
+        tf.debugging.assert_shapes([at, 'ND'])
+        mean, cov = model.predict(at)   # [N, L], [N, L]
+        return mean + tf.math.sqrt(cov) * eps[:, None, :]  # [S, N, L]
+
+    return sample
+
+
+# todo the batch size problem
+def batch_latent_sampler_with_reparametrisation(
+    dataset: Dataset,
+    model: ProbabilisticModel,
+    sample_size: int,
+    batch_size: int,
+    observation_shape: ShapeLike,
+    jitter: Union[tf.Tensor, float] = DEFAULTS.JITTER
+) -> Callable[[tf.Tensor], tf.Tensor]:
+    """
+    Returns batch-correlated samples according to the reparametrization trick.
+    The sample size S is determined by eps,
+    the batch size B must be compatible between eps and at,
+    N is the number of batches of query points
+    and L is the number of latent processes.
+    :param dataset:
+    :param model:
+    :param sample_size:
+    :param batch_size:
+    :param observation_shape:
+    :param jitter:
+    :return: a [S, N, B, L] tensor
+    """
+    eps = tf.random.normal(
+        observation_shape + [batch_size, sample_size], dtype=dataset.observations.dtype
+    )  # [L, B, S]
+
+    def sample(at: tf.Tensor) -> tf.Tensor:
+        # todo do we need this shape check?
+        tf.debugging.assert_shapes([(at, [..., sample_size, batch_size, '_'])])
+
+        # todo check this is the same as what victor wrote
+        mean, cov = model.predict(at, full_cov=True)  # [N, B, L], [N, L, B, B]
+        identity = tf.eye(batch_size, dtype=cov.dtype)  # [B, B]
+        cov_cholesky = tf.linalg.cholesky(cov + jitter * identity)  # [N, L, B, B]
+        return mean + tf.transpose(cov_cholesky @ eps, [3, 0, 2, 1])  # [S, N, B, L]
```

Comment:
I noticed you changed the computations here. Are you certain the math is still right? the previous version was taken from gpflow